ubirch/ubirch_protocol.py

```
@@ -172,43 +264,91 @@ def message_chained(self, uuid: UUID, type: int, payload: any) -> bytes:
         """
-        return self._verify(uuid, self._hash(message), signature)
+        # check the lower four bits of the version byte
+        lowerFour = versionByte & 0x0f
 
-    def message_verify(self, message: bytes) -> list:
+        if lowerFour == 0x01:
+            return UNPACKED_UNSIGNED_UPP_INDEX_TABLE[targetField]
+        elif lowerFour == 0x02:
+            return UNPACKED_SIGNED_UPP_INDEX_TABLE[targetField]
+        elif lowerFour == 0x03:
+            return UNPACKED_CHAINED_UPP_INDEX_TABLE[targetField]
+        else:
+            # unknown lower four bits; error
+            raise ValueError("Invalid lower four bits of the UPP version byte: %s" % bin(lowerFour))
+
+    def upp_msgpack_split_signature(self, msgpackUPP) -> (bytes, bytes):
+        """!
+        Separate the signature from the msgpack
+        @param msgpackUPP The msgpack encoded upp
+        @return A tuple consisting of the message without the signature and the signature
         """
-        Verify the integrity of the message and decode the contents.
-        Throws an exception if the message is not verifiable.
-        :param message: the msgpack encoded message
-        :return: the decoded message
+        try:
+            if msgpackUPP[1] >> 4 == 2:
+                # version 2 upp - 2 byte signature header
+                return (msgpackUPP[:-66], msgpackUPP[-64:])
+            elif msgpackUPP[1] >> 4 == 1:
+                # version 1 upp - 3 byte signature header
+                return (msgpackUPP[:-67], msgpackUPP[-64:])
+            else:
+                raise ValueError("Invalid UPP version byte: %02x" % msgpack[1])
+        except IndexError:
+            raise ValueError("The UPP-msgpack is too short: %d bytes" % len(msgpackUPP))
+
+    def verify_signature(self, uuid: UUID, msgpackUPP: bytes) -> bool:
+        """!
+        Verify the integrity of the message and decode the contents
+        Raises an value error when the message is too short
```

Also raises ValueError on other errors. In my opinion, it should also raise an exception if the message is not verifiable instead of returning bool -> see below comment.

---

updated to raise a ValueError